cpp/src/arrow/compute/api_scalar.h

```diff
@@ -278,6 +278,23 @@ struct ARROW_EXPORT DayOfWeekOptions : public FunctionOptions {
   uint32_t week_start;
 };
 
+struct ARROW_EXPORT WeekOptions : public FunctionOptions {
+ public:
+  explicit WeekOptions(bool week_starts_monday = true, bool count_from_zero = false,
+                       bool first_week_in_year = false);
+  constexpr static char const kTypeName[] = "WeekOptions";
+  static WeekOptions Defaults() { return WeekOptions{}; }
+
+  /// What day does the week start with (Monday=true, Sunday=false)
+  bool week_starts_monday;
+  /// Dates from current year that fall into last ISO week of the previous year return
+  /// 0 if true and 52 or 53 if false.
+  bool count_from_zero;
```

Comment:
Similar question here vs. `DayOfWeekOptions::one_based_numbering`.

Reply:
There's a difference between the two - `one_based_numbering` controls if Monday=1 or Monday=0.
`count_from_zero` defines what is the last week of previous year that is still in this year - 0 or 52.

Reply:
Er... then what is the difference between `count_from_zero` and `first_week_is_fully_in_year`?

Reply:
`first_week_is_fully_in_year` controls if the whole week (from Sunday or Monday) is in new year or only most of it (4 days).

Reply:
See here for all the possibilities: https://github.com/apache/arrow/pull/11026#issuecomment-910649581

Reply:
That table was taken [from MySQL](https://dev.mysql.com/doc/refman/8.0/en/date-and-time-functions.html#function_week)

Reply:
But why do we have to support all this? We are not trying to reimplement MySQL. And I don't see any such flexibility in PostgreSQL or SQL Server.

Reply:
Well lubridate supports a subset of this (https://lubridate.tidyverse.org/reference/week.html). Pandas only really supports iso_week.
Perhaps @ianmcook can tweigh in on this.

Reply:
"week() returns the number of complete seven day periods that have occurred between the date and January 1st, plus one" : that doesn't sound like what you're implementing, does it?

Reply:
It could mean mode 7 (if you read complete as all days in the week starting with Monday).

Reply:
You're right though ([source](https://rdrr.io/cran/lubridate/src/R/accessors-week.r):
```
week <- function(x)
  (yday(x) - 1) %/% 7 + 1

#' @rdname week
#' @export
"week<-" <- function(x, value)
  x <- x + days((value - week(x)) * 7)

.other_week <- function(x, week_start) {
  x <- as.POSIXlt(x)
  date <- make_date(year(x), month(x), day(x))
  wday <- wday(x, week_start = week_start)
  date <- date + (4 - wday)
  jan1 <- as.numeric(make_date(year(date), 1, 1))
  1L + (as.numeric(date) - jan1) %/% 7L
}

#' @description `isoweek()` returns the week as it would appear in the ISO 8601
#'   system, which uses a reoccurring leap week.
#' @rdname week
#' @export
isoweek <- function(x) {
  .other_week(x, 1)
}

#' @description `epiweek()` is the US CDC version of epidemiological week. It
#'   follows same rules as `isoweek()` but starts on Sunday. In other parts of
#'   the world the convention is to start epidemiological weeks on Monday,
#'   which is the same as `isoweek`.
#'
#' @rdname week
#' @export
epiweek <- function(x) {
  .other_week(x, 7)
}
```

Reply:
Ha. Well, that sounds like a useful function :-)

Reply:
It's great.

Let's try to settle on functionality we need here then.

There's the 8 modes of MySQL (they include the ISO week and epiweek) and yday/7. Which of these 9 do we want/need?

Reply:
SQL server allows you to choose the starting day of the week (also, "January 1 of any year defines the starting number for the week datepart"):
https://docs.microsoft.com/en-us/sql/t-sql/functions/datepart-transact-sql?view=sql-server-ver15#week-and-weekday-datepart-arguments

PostgreSQL only seems to support ISO week numbers: https://www.postgresql.org/docs/current/functions-datetime.html#FUNCTIONS-DATETIME-EXTRACT

[Arrow](https://arrow.readthedocs.io/en/latest/) (the other one) does not seem to support non-ISO week numbers.

Reply:
Note that lubridate's "epiweek" is simply an [American week number](https://savvytime.com/week-number/united-states/2021) as far as I understand.

So we could have a "week" function with configurable week start day, or more simply a "us_week" function with weeks starting on Sunday (both with semantics otherwise similar to "iso_week")

Reply:
Or of course we have a "week" function allowing to configure the starting day of the week. Week number 1 is then the week containing the day `start of week + 3` (e.g. Thursday for weeks starting on Monday, Wednesday for weeks starting on Sunday...).

Reply:
My understanding is that supporting all the MySQL modes is fairly straightforward and provides the greatest flexibility. Using these three bool options to specify the modes seems like a clever way to do it.

I do not think it is important to provide a mode here that computes `day_of_year() / 7`. If someone wants that they can do `day_of_year() / 7`. We can add a binding for the lubridate `week()` function that does that.

Reply:
While the implementation is reasonably simple, my main concern here is a head-scratching API. Supporting the MySQL modes does not sound like an interesting goal in itself, and exposing obscure parameters makes the API more difficult to learn and use.

We should look for actual cases. One actual use case is US week numbering (like `epiweek` in lubridate). Are there any others?

Reply:
From [MySQL docs](https://dev.mysql.com/doc/refman/8.0/en/date-and-time-functions.html#function_week) about why the 8 modes:

> If a date falls in the last week of the previous year, MySQL returns 0 if you do not use 2, 3, 6, or 7 as the optional mode argument.

> One might argue that WEEK() should return 52 because the given date actually occurs in the 52nd week of 1999. WEEK() returns 0 instead so that the return value is “the week number in the given year.” This makes use of the WEEK() function reliable when combined with other functions that extract a date part from a date.

> If you prefer a result evaluated with respect to the year that contains the first day of the week for the given date, use 0, 2, 5, or 7 as the optional mode argument.

Reply:
ping

Reply:
If the concern is primarily a confusing API, could we offer default options for 'common' modes like us_week and iso_week? Then the individual options could be left up to people who need/understand what they want.